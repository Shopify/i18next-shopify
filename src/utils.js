/* eslint-disable prettier/prettier */
// Dynamically & Optionally import React.
let React;
// eslint-disable-next-line promise/catch-or-return
import('react')
  .then((module) => {
    React = module;
  })

const arr = [];
const each = arr.forEach;

// Copied from https://github.com/i18next/i18next-icu/blob/370027c829e240b36b2f6e5d648be35453c9e6d8/src/utils.js
export function defaults(obj, ...args) {
  each.call(args, (source) => {
    if (source) {
      for (const prop in source) {
        if (obj[prop] === undefined) {
          obj[prop] = source[prop];
        }
      }
    }
  });
  return obj;
}

/**
 * Given a value that may contain interpolated values, replaces all occurrences of the specified text
 * with the specified replacement value, and returns a new value with the replacements made.
 * If React is available, this function supports nested interpolated values, allowing you to replace
 * text with React elements or arrays of React elements.
 *
 * @param {string|object|Array} interpolated - The value to replace occurrences of the specified text in.
 * @param {string|RegExp} find - The text or regular expression to search for in the interpolated value.
 * @param {string|object|Array} replace - The value to replace occurrences of the specified text with.
 * @returns {string|object|Array} A new value with the specified text replaced.
 */
export function replaceValue(interpolated, find, replace) {
  if (!React) {
    return interpolated.replace(find, replace);
  }

  switch (typeof interpolated) {
    case 'string': {
      const split = interpolated.split(find);

      if (split.length !== 1 && typeof replace === 'object') {
        // Replace is an object. Return a React fragment with the replacement.

        return (
          <React.Fragment key={split[0]}>
            {split[0]}
            {replace}
            {split[1]}
          </React.Fragment>
        );
      }

      // This is a simple text replacement.
      return interpolated.replace(find, replace);
    }

    case 'object':
      if (Array.isArray(interpolated)) {
        // The interpolated element is an array, call replaceValue on each item
        return interpolated.map((item) => replaceValue(item, find, replace)).flat();
      }

      // The interpolated element is an object with props, check its children
      if (interpolated?.props) {
        let hasChanged = false;

        const newChildren = replaceValue(interpolated.props.children, find, replace);
        if (newChildren !== interpolated.props.children) {
          hasChanged = true;
        }

        return hasChanged
          ? React.cloneElement(interpolated, {children: newChildren})
          : interpolated;
      }
  }

  // The interpolated element is something else, just return it
  return interpolated;
}
