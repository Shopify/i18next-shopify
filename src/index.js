import * as utils from "./utils.js";

function getDefaults() {
  return {};
}

class ShopifyFormat {
  constructor(options) {
    this.type = "i18nFormat";

    this.init(null, options);
  }

  init(i18next, options) {
    const i18nextOptions =
      (i18next && i18next.options && i18next.options.i18nFormat) || {};
    this.options = utils.defaults(
      i18nextOptions,
      options,
      this.options || {},
      getDefaults()
    );
    this.i18next = i18next;
  }

  // i18next is roughly a superset of Shopify's format.
  // TODO: Implement this to make sure that we enforce just the subset
  // TODO: Implement this to make ordinal pluralization work

  // Implement custom interpolation logic
  // While i18next and Shopify's format both use the mustache syntax for interpolation,
  // Shopify uses the `ordinal` interpolation for ordinal pluralization, while i18next uses `count`.
  parse(res, options, lng, ns, key, info) {
    const hadSuccessfulLookup = info && info.resolved && info.resolved.res;

    // Interpolations
    const MUSTACHE_FORMAT = /{{\s*(\w+)\s*}}/g;
    const matches = res.match(MUSTACHE_FORMAT);
    if (!matches) return res;
    matches.forEach((match) => {
      const interpolation_key = match.replace(MUSTACHE_FORMAT, "$1");
      const value =
        interpolation_key === "ordinal"
          ? options["count"] || options["ordinal"]
          : options[interpolation_key];
      if (value !== undefined) {
        res = res.replace(match, value);
      }
    });
    return res;
  }

  // Add any other locations that should be searched first for an answer to the lookup
  // Add keys to `finalKeys` in reverse order (e.g., least specific -> most specific)
  // Useful when defining keys for pluralization or other context cases (e.g., grammatical gender)
  addLookupKeys(finalKeys, key, code, ns, options) {
    const needsPluralHandling =
      (options.count !== undefined && typeof options.count !== "string") ||
      typeof options.ordinal == "number";

    if (needsPluralHandling) {
      if (!this.i18next.translator.pluralResolver.shouldUseIntlApi()) {
        throw new Error("We need that to exist"); // TODO
      }

      // Shopify uses the "ordinal" interpolation for ordinal pluralization (i.e., {{ordinal}}), users will expect to
      // do lookups with `i18n.t("key", { ordinal: 1 })`.
      // However, the i18next pluralization system uses the "count" option for both cardinal and ordinal pluralization
      // so users will expect to do lookups with `i18n.t("key", { count: 1, ordinal: true })`.
      // So we support either, using count if provided.
      // There is an edge case: if `ordinal` were set explicitly to 0, and `count` is provided, we behave as i18next
      // does, treating it as cardinal pluralization.
      const needsOrdinalHandling =
        options.ordinal ||
        (options.ordinal == 0 && options.count === undefined);

      const pluralRule = this.i18next.translator.pluralResolver.getRule(code, {
        ...options,
        ordinal: needsOrdinalHandling,
      });

      if (needsOrdinalHandling) {
        const ruleName = pluralRule.select(
          options.count !== undefined ? options.count : options.ordinal
        );
        const pluralSuffix = `${this.i18next.options.keySeparator}ordinal${this.i18next.options.keySeparator}${ruleName}`;
        finalKeys.push(key + pluralSuffix);
      } else {
        const ruleName = pluralRule.select(options.count);

        // Fallback to "other" key
        if (ruleName !== "other") {
          const otherSubkey = `${this.i18next.options.keySeparator}other`;
          finalKeys.push(key + otherSubkey);
        }

        // Pluralization rule key
        const pluralSuffix = `${this.i18next.options.keySeparator}${ruleName}`;
        finalKeys.push(key + pluralSuffix);

        // Explicit "0" and "1" keys
        if (options.count === 0) {
          const explicit0Subkey = `${this.i18next.options.keySeparator}0`;
          finalKeys.push(key + explicit0Subkey);
        } else if (options.count === 1) {
          const explicit1Subkey = `${this.i18next.options.keySeparator}1`;
          finalKeys.push(key + explicit1Subkey);
        }
      }
    }

    return finalKeys;
  }

  getResource(lng, _ns, key, options) {
    // This is a copy of
    // https://github.com/i18next/i18next/blob/ac4b6701c3ce9596e4eb88f5d774ca66f05d71fb/src/ResourceStore.js#L34-L56
    // except we ignore the namespace, since Shopify's format doesn't have namespaces.
    const keySeparator =
      options.keySeparator !== undefined
        ? options.keySeparator
        : this.i18next.translator.resourceStore.options.keySeparator;

    const ignoreJSONStructure =
      options.ignoreJSONStructure !== undefined
        ? options.ignoreJSONStructure
        : this.i18next.translator.resourceStore.options.ignoreJSONStructure;

    let path = [lng];
    if (key && typeof key !== "string") path = path.concat(key);
    if (key && typeof key === "string")
      path = path.concat(keySeparator ? key.split(keySeparator) : key);

    if (lng.indexOf(".") > -1) {
      path = lng.split(".");
    }

    const result = utils.getPath(
      this.i18next.translator.resourceStore.data,
      path
    );
    if (result || !ignoreJSONStructure || typeof key !== "string")
      return result;

    return utils.deepFind(
      this.i18next.translator.resourceStore.data[lng],
      key,
      keySeparator
    );
  }
}

ShopifyFormat.type = "i18nFormat";

export default ShopifyFormat;
