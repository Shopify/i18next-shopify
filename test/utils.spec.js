import React from 'react';

import {replaceValue} from '../src/utils';

describe('Utils', () => {
  describe('replaceValue', () => {
    it('replaces a string with a single level of interpolation', () => {
      expect(replaceValue('Hello, {{name}}!', '{{name}}', 'John')).toBe(
        'Hello, John!',
      );
    });

    it('replaces a string within an array with a single level of interpolation', () => {
      expect(
        replaceValue(['Blah! ', 'Hello, {{name}}!'], '{{name}}', 'John'),
      ).toStrictEqual(['Blah! ', 'Hello, John!']);
    });

    it('replaces a string with React elements', () => {
      expect(
        replaceValue('Hello, {{name}}!', '{{name}}', <span>John</span>),
      ).toMatchSnapshot();
    });

    it('returns the original string when there is no match with given React elements', () => {
      expect(
        replaceValue('Hello, {{name}}!', '{{no_match}}', <span>John</span>),
      ).toBe('Hello, {{name}}!');
    });

    it('replaces a string within nested React elements', () => {
      expect(
        replaceValue(
          <>
            Hello there <span>{'{{name}}'}</span>
          </>,
          '{{name}}',
          'John',
        ),
      ).toMatchSnapshot();
    });

    it('replaces a string with a regular expression', () => {
      expect(replaceValue('Hello, {{name}}!', /{{name}}/, 'John')).toBe(
        'Hello, John!',
      );
    });
  });
});
