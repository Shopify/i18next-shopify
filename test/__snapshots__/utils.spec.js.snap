// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils replaceValue replaces a string with React elements 1`] = `
<React.Fragment>
  Hello, 
  <span>
    John
  </span>
  !
</React.Fragment>
`;

exports[`Utils replaceValue replaces a string within React elements 1`] = `
<React.Fragment>
  Hello there 
  <span>
    John
  </span>
</React.Fragment>
`;

exports[`Utils replaceValue replaces a string within nested React elements 1`] = `
<React.Fragment>
  Hello there 
  <span>
    John
  </span>
</React.Fragment>
`;
